crates/bevy_ptr/README.md

```diff
@@ -6,8 +6,104 @@
+|:------------------|:----------|:------|:-------------|:------|:-------|:---------------|:-----------------|
+|`Box<T>`           |Owned      |Yes    |Yes           |Yes    |Yes     |Yes             |Yes               |
+|`&'a mut T`        |Yes        |Yes    |Yes           |Yes    |Yes     |Yes             |Yes               |
+|`&'a T`            |Yes        |No     |Yes           |Yes    |Yes     |No              |Yes               |
+|`&'a UnsafeCell<T>`|Yes        |Maybe  |Yes           |Yes    |Yes     |Yes             |Yes               |
+|`NonNull<T>`       |No         |Yes    |Yes           |No     |Yes     |No              |No                |
+|`*const T`         |No         |No     |Yes           |No     |No      |No              |No                |
+|`*mut T`           |No         |Yes    |Yes           |No     |No      |No              |No                |
+|`*const ()`        |No         |No     |No            |No     |No      |No              |No                |
+|`*mut ()`          |No         |Yes    |No            |No     |No      |No              |No                |
+
+`&T`, `&mut T`, and `Box<T>` are by far the most common pointer types that Rust developers will see. They're the only ones in this list that are entirely usable
+without the use of `unsafe`.
+
+`&UnsafeCell<T>` is the first step away from safety. `UnsafeCell` is the *only* way to get a mutable borrow from an immutable one in the language, so it's the
+base primitive for all interior mutability in the language: `Cell<T>`, `RefCell<T>`, `Mutex<T>`, `RwLock<T>`, etc. are all built on top of
+`UnsafeCell<T>`. To safety convert `&UnsafeCell<T>` into a `&T` or `&mut T`, the caller must guarantee that all simultaneous access follow Rust's aliasing rules.
+
+`NonNull<T>` takes quite a step down from the aforementioned types. In addition to allowing aliasing, it's the first pointer type on this list to drop both
+lifetimes and the alignment guarantees of borrows. The only guarantees it provides are that the pointer is not null, and that it points to a valid instance
+of type `T`. If you've ever worked with C++, `NonNull<T>` is very close to a C++ reference (`T&`).
+
+`*const T` and `*mut T` are what most developers with a background in C or C++ would consider pointers.
+
+`*const ()` is the bottom of this list. They're the Rust equivalent to C's `void*`.  Note that Rust doesn't formally have a concept of type that holds an arbitrary
+untyped memory address. Pointing at the unit type (or some other zero-sized type) just happens to be the convention here. They show up occasionally when dealing
+with FFI and the rare occasion where dynamic dispatch is required, but a trait is too constraining of an interface to work with. A great example of this are the
+[RawWaker] APIs, where a singular trait (or set of traits) may be insufficient to capture all usage patterns. `*mut ()`, as far as I know, should never be used.
```

Comment:
Using first person is a bit weird in module docs like this.

Reply:
I copied this over from a personal blog post draft, whoops.

Reply:
Done.